src/main/java/com/minecolonies/coremod/inventory/CraftingGUIBuilding.java

```diff
@@ -0,0 +1,314 @@
+                    {
+                        return false;
+                    }
+                });
+            }
+        }
+
+        int i;
+        for (i = 0; i < PLAYER_INVENTORY_ROWS; i++)
+        {
+            for (int j = 0; j < PLAYER_INVENTORY_COLUMNS; j++)
+            {
+                addSlotToContainer(new Slot(
+                        playerInventory,
+                        j + i * PLAYER_INVENTORY_COLUMNS + PLAYER_INVENTORY_COLUMNS,
+                        PLAYER_INVENTORY_INITIAL_X_OFFSET + j * INVENTORY_OFFSET_EACH,
+                        PLAYER_INVENTORY_INITIAL_Y_OFFSET + i * INVENTORY_OFFSET_EACH
+                ));
+            }
+        }
+
+        for (i = 0; i < PLAYER_INVENTORY_COLUMNS; i++)
+        {
+            addSlotToContainer(new Slot(
+                    playerInventory, i,
+                    PLAYER_INVENTORY_INITIAL_X_OFFSET + i * INVENTORY_OFFSET_EACH,
+                    PLAYER_INVENTORY_HOTBAR_OFFSET
+            ));
+        }
+
+        this.onCraftMatrixChanged(this.craftMatrix);
+    }
+
+    @Override
+    protected final Slot addSlotToContainer(final Slot slotToAdd)
+    {
+        return super.addSlotToContainer(slotToAdd);
+    }
+
+    /**
+     * Callback for when the crafting matrix is changed.
+     */
+    @Override
+    public void onCraftMatrixChanged(final IInventory inventoryIn)
+    {
+        super.onCraftMatrixChanged(inventoryIn);
+        this.craftResult.setInventorySlotContents(0, CraftingManager.getInstance().findMatchingRecipe(this.craftMatrix, this.worldObj));
+    }
+
+    @Override
+    public boolean canInteractWith(final EntityPlayer playerIn)
+    {
+        return true;
+    }
+
+    @Override
+    public ItemStack slotClick(int slotId, int clickedButton, ClickType mode, EntityPlayer playerIn)
+    {
+        if (slotId >= 1 && slotId < 5)
+        {
+            if (mode == ClickType.PICKUP || mode == ClickType.PICKUP_ALL ||
+                  mode == ClickType.SWAP) // 1 is shift-click
+            {
+                Slot slot = this.inventorySlots.get(slotId);
+
+                ItemStack dropping = playerIn.inventory.getItemStack();
+
+                return handleSlotClick(slot, dropping);
+            }
+
+            return ItemStack.EMPTY;
+        }
+
+        return super.slotClick(slotId, clickedButton, mode, playerIn);
+    }
+
+    public ItemStack handleSlotClick(Slot slot, ItemStack stack)
+    {
+        if (stack.getCount() > 0)
+        {
+            ItemStack copy = stack.copy();
+            copy.setCount(1);
+            slot.putStack(copy);
+        }
+        else if (slot.getStack().getCount() > 0)
+        {
+            slot.putStack(ItemStack.EMPTY);
+        }
+
+        return slot.getStack().copy();
+    }
+
+    @Nullable
+    @Override
+    public ItemStack transferStackInSlot(final EntityPlayer playerIn, final int index)
+    {
+        if (index <= 5)
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Assign this magic number 5 to a well-named constant, and use the constant instead. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)